src/backend/gl/src/device.rs

```diff
@@ -1568,26 +1550,100 @@ impl d::Device<B> for Device {
         fence: &n::Fence,
         timeout_ns: u64,
     ) -> Result<bool, d::OomOrDeviceLost> {
-        if !self.share.private_caps.sync {
-            return Ok(true);
+        // TODO:
+        // This can be called by multiple objects wanting to ensure they have exclusive
+        // access to a resource. How much does this call costs ? The status of the fence
+        // could be cached to avoid calling this more than once (in core or in the backend ?).
+        let gl = &self.share.context;
+        match fence.0.get() {
+            n::FenceInner::Idle { signaled } => {
+                if !signaled {
+                    warn!("Fence ptr {:?} is not pending, waiting not possible", fence.0.as_ptr());
+                }
+                Ok(signaled)
+            }
+            n::FenceInner::Pending(Some(sync)) => {
+                // TODO: Could `wait_sync` be used here instead?
+                match gl.client_wait_sync(
+                    sync,
+                    glow::SYNC_FLUSH_COMMANDS_BIT,
+                    timeout_ns as i32,
+                ) {
+                    glow::TIMEOUT_EXPIRED => Ok(false),
+                    glow::WAIT_FAILED => {
+                        if let Err(err) = self.share.check() {
+                            error!("Error when waiting on fence: {:?}", err);
+                        }
+                        Ok(false)
+                    }
+                    _ => Ok(true),
```

Comment:
Would it be faster / more correct here to transition the fence here to `FenceStatus::Idle { true }` to avoid future `glClientWaitSync` calls?  The metal backend did not do this so I also didn't do it.

Reply:
we should replace `_` by a real arm, and leave `_` for unexpected stuff. And yes, we should change it to `Idle { signaled: true }` both here and in Metal

Reply:
That's from previosly, but yeah I agree and I will change it.